### ISSUE TITLE
Can't use asset() Twig function with Vite dev server due to missing manifest.json

### ISSUE THREAD

**Issue by WebMechanic** [lhapaipai/vite-plugin-symfony#8]
Salût,

Vite's dev mode doesn't create a `manifest.json` which is required to use Symfony's [`asset()` function](https://symfony.com/doc/current/reference/twig_reference.html#asset). 
Symfony Encore did create that file alongside `entrypoints.json` in both dev and production (build) mode.

Using it will cause Symfony to throw:
> An exception has been thrown during the rendering of a template ("Asset manifest file "<project-path>/public/build/manifest.json" does not exist. Did you forget to build the assets with npm or yarn?").

It's a fine and good after `yarn build`, but I also need access to `{{ asset(...) }}` in dev mode to get files from our media bucket.

Thanks.

**Comment by WebMechanic** [comment]
It appears to be sufficent for the _file to exist with an empty JSON literal `{}`_ inside to satisfy Symfony. 
If the Plugin detects dev mode it should write a manifest placholder.

**However:** this is only tested on localhost w/o any fancy asset packages that eventually create CDN URLs:
local file: `/public/images/hero.webp`
Twig: `{{ asset('images/hero.webp') }}`

Other asset packages may do more elaborate checks on the JSON data 🤷‍♂️ 

Thanks

**Comment by lhapaipai** [comment]
hi @WebMechanic ,
have you tried the the vite asset component described in the https://github.com/lhapaipai/vite-bundle#symfony-asset-component ?
vite does not generate a manifest file during the development phase but it is possible to request its server to retrieve the assets, and this is what 'Pentatrion\ViteBundle\Asset\ViteAssetVersionStrategy' does.

**Comment by WebMechanic** [comment]
Salút @lhapaipai,

allright, so _that's_ what this was for! Admittedly I only glanced over that section after the last release and forgot about it when I eventually used asset() in my little local test project.

It _partially_ solved the problem in `dev` mode.

There's also a (default?) `config/packages/assets.yaml` present which defines `json_manifest_path`. Sumply enabling ViteAssetVersionStrategy leads to another Symfony error:
> Invalid configuration for path "framework.assets": You cannot use both "version_strategy" and "json_manifest_path" at the same time under "assets".

To get rid of that error I had to overwrite `json_manifest_path` in `framework.yaml`
```yaml
# config/packages/framework.yaml
framework:
    # Vite dev mode assets() support
    assets:
        version_strategy: 'Pentatrion\ViteBundle\Asset\ViteAssetVersionStrategy'
        # undo assets standard manifest resolution
        json_manifest_path: null
```

This doesn't seem right either, as it's "unconditional": no distinction of dev vs. production on _Symfony's end_. 
Flipping between Vite's dev/build modes and changing these yamls at some point there was no `entrypoints.json` when in dev mode...

Maybe it's the mxiture of both modes on the same machine using only _one_ webserver, _one_ node/vite environment with the _same_ "DocumentRoot" /public.

I need to work on some other project now, but I keep digging and would like to keep this issue open. 
I'll eventually document my findings in a discussion post. Somehow there's missing a pience in the whole plot.

Merci et bonne journée

**Comment by lhapaipai** [comment]
Hi @WebMechanic,
It's not an error but normal behavior !!! by defining a custom strategy you delegate the resolution of your assets to `ViteAssetVersionStrategy`  which should take care of everything ! there is no `json_manifest_path`. look at this : https://github.com/lhapaipai/vite-bundle/blob/b48303f96f745bfa99166128cbf39b16d2319996/src/Resources/config/services.yaml#L47
there is no entrypoints.json in dev mode and it's not a problem for `ViteAssetVersionStrategy` , it doesn't complain about it
https://github.com/lhapaipai/vite-bundle/blob/b48303f96f745bfa99166128cbf39b16d2319996/src/Asset/ViteAssetVersionStrategy.php#L65
and should return a path like `http://127.0.0.1:5174/build/path/to/your/asset.jpg` in dev mode.
**You can use this asset() function only with assets referenced by JavaScript or CSS files.**
otherwise look at the recipe at the end with the glob : https://github.com/lhapaipai/vite-bundle#symfony-asset-component

please could you post your bundle related issues in : https://github.com/lhapaipai/vite-bundle, here it is only related to the vite plugin.

Finally if you encounter problems of understanding could you rather open discussions rather than issues.
Thank you very much and good evening !

**Comment by WebMechanic** [comment]
je suis désolé.

> please could you post your bundle related issues in : https://github.com/lhapaipai/vite-bundle, here it is only related to the vite plugin.

The initial **issue** was Symfony complaining about a missing file _generated by the Plugin_. 
So I came here to the _Plugin_ issue tracker.

Both bundle and plugin are massivly intertwined and rely on each other's settings, outputs, and behaviour that all change with the diffent modes, dev or build.
As a _user_, it's hard to tell which part may cause trouble on the other side in any of these modes. What works smooth in one suddenly makes the other side complain, as does switching modes.

Your proposal to use the `ViteAssetVersionStrategy` solved one bit, but caused the follow-up **issue** with settings from the standard assets.yaml that _does_ contain a `json_manifest_path` - a setting I did not create, but obviously active at runtime anyway -- so now we're apparently back in Symfony territory. I got rid of that new error be adding `json_manifest_path:null` as explained.

> look at this : https://github.com/lhapaipai/vite-bundle/blob/b48303f96f745bfa99166128cbf39b16d2319996/src/Resources/config/services.yaml#L47

what I see is a hardcoded `/public` folder that's a configuable value in several places (and thus could change), and should probably refer to `%pentatrion_vite.public_dir%` instead? You want me to open an issue for that in [vite-bundle](https://github.com/lhapaipai/vite-bundle)?

> could you rather open discussions rather than issues.

will do, and I'll illustrate my setup of that Gordian Knot -- but it'll probably be in the "wrong" repo anyway 'cos it involves both sides.

Merci beaucoup et bonne journée.

**Comment by lhapaipai** [comment]
Hi @WebMechanic, 
thanks for your hardcoded /public folder issue. I just solved it in v2.2.2

Interesting your `json_manifest_path`, probably a previously config value initiated previously in your project ?
because it seems to me that its default value is null.
```bash
./bin/console debug:config framework
```
```yaml
framework:
    ...
    assets:
        enabled: true
        strict_mode: false
        version_strategy: null
        version: null
        version_format: '%%s?%%s'
        json_manifest_path: null
        base_path: ''
        base_urls: {  }
        packages: {  }
...
```
sorry for my rudeness in the previous message.
Bon après-midi !

**Comment by WebMechanic** [comment]
Salût @lhapaipai ,

> Interesting your `json_manifest_path`, probably a config value initiated previously in your project ? 

I frankly wouldn't know. 
That whole `assets.yaml` looks like some standard file to me created by Symfony or some Flex recipe. I haven't touched it and found this value digging for a probably cause related to any of the troubles I had with these two .json files.

This is all from a plain vanilla project created by the latest Symfony binary only to play with your bundle and the plugin.

I only edited settings related to the "public" and "build" directories and URLs back and forth 'cos they'll eventually differ and must be changed to match the actual projects that are due to relaunch and I want to get rid of Encore and Webpack for good.

> sorry for my rudeness in the previous message.

never mind, I know how "naive questions" can be a pita for those who know the code and believe it must all be so obvious. 
TL;DR it almost never is for the _user_.
Most certainly not with systems as complex as Symfony and bundling.

I'll install another fresh project this weekend and try to "nudge" the setting towards what the final environment would be for the real projects, which are very much more complex.

Bon week-end !
